### PR TITLE
FILE-411 Redirection feature added

### PR DESCRIPTION
**Pull request by kastoestoramadus**
tests are passing, to be manually tested. Config of allowed domains need to be added.
Review may start.

### CODE REVIEW

File: app/uk/gov/hmrc/fileupload/controllers/FileUploadController.scala [hunk 1]
```diff
@@ -39,12 +41,22 @@ class FileUploadController(withValidEnvelope: WithValidEnvelope,
                            now: () => Long)
                           (implicit executionContext: ExecutionContext) extends Controller {
 
-  def uploadWithEnvelopeValidation(envelopeId: EnvelopeId, fileId: FileId) =
+  val helper = new RedirectionFeature(Seq()) // FIXME configuration
```
Comment (kings13y) [inline app/uk/gov/hmrc/fileupload/controllers/FileUploadController.scala]: Prob a minor thing. But not a fan of having generically named val's called things like _helper_ esp if it's a specific to a feature of the class.

File: app/uk/gov/hmrc/fileupload/controllers/RedirectionFeature.scala [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+import scala.util.{Failure, Success, Try}
+
+
+class RedirectionFeature(allowedHosts: Seq[String]) {
+  import EnvelopeChecker.logAndReturn
+
+  case class ValidatedUrl(url: String)
+  case class RedirectUrlsO(succ: Option[ValidatedUrl], fail: Option[ValidatedUrl])
+
+  def redirect(successUrlO: Option[String], failureUrlO: Option[String])
```
Comment (kings13y) [inline app/uk/gov/hmrc/fileupload/controllers/RedirectionFeature.scala]: how come the vals in the sig have the '0' appended on ?

File: app/uk/gov/hmrc/fileupload/controllers/RedirectionFeature.scala [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+      val suspect = new URL(url)
+      if(suspect.getProtocol != "https")
+        throw new MalformedURLException("Https is required.")
+      if(!allowedHosts.contains(suspect.getHost))
+        throw new MalformedURLException("Given domain is not allowed.")
+    }.map( _ => ValidatedUrl(
+    url.takeWhile( c => !(c=='?' || c=='#'))
+    ))
+  }
+  // with cats it should be much better:
```
Comment (kings13y) [inline app/uk/gov/hmrc/fileupload/controllers/RedirectionFeature.scala]: No problem with using Cats fwiw. Looks like you just want to use an Applicative here, no ?

File: app/uk/gov/hmrc/fileupload/controllers/RedirectionFeature.scala (hunk 2 above)
Comment (kastoestoramadus) [inline app/uk/gov/hmrc/fileupload/controllers/RedirectionFeature.scala]: O like Option, T like Try, F like Future

File: app/uk/gov/hmrc/fileupload/controllers/RedirectionFeature.scala (hunk 3 above)
Comment (kastoestoramadus) [inline app/uk/gov/hmrc/fileupload/controllers/RedirectionFeature.scala]: I am not yet familiar with cats. I wanted to write it fast.

Comment (kings13y) [inline app/uk/gov/hmrc/fileupload/controllers/RedirectionFeature.scala]: OK. Talking to Peter P, will setup a hack session on it fwiw. Nps.

File: app/uk/gov/hmrc/fileupload/controllers/RedirectionFeature.scala (hunk 2 above)
Comment (kings13y) [inline app/uk/gov/hmrc/fileupload/controllers/RedirectionFeature.scala]: Not seen that as a syntactic convention before. Prob not nec imho (type's tell you it's an option). I'm not hung up on this though tbh.

**Comment by kings13y** [comment]
@kastoestoramadus Waiting for more reviews ? otherwise, I'm OK to merge this PR.

File: app/uk/gov/hmrc/fileupload/controllers/RedirectionFeature.scala (hunk 2 above)
Comment (kastoestoramadus) [inline app/uk/gov/hmrc/fileupload/controllers/RedirectionFeature.scala]: if the type is not given explicitly then without Intelij it is hard to get the type, f.ex. at github reviews.
This PR code might be better without need of such naming but no time for that.

**Comment by kastoestoramadus** [comment]
@YuanZHANG910 @JTY009 @codesurf42

File: app/uk/gov/hmrc/fileupload/controllers/RedirectionFeature.scala (hunk 2 above)
Comment (kings13y) [inline app/uk/gov/hmrc/fileupload/controllers/RedirectionFeature.scala]: ? Might be at cross purposes. I prefer having the type there​ explicitly. Just the variable naming convention I queried.

**Pull request merged**